### PR TITLE
Feat: Stat points incrementation from trophies/prestige etc

### PR DESCRIPTION
**Pull request by Andrew-Gan**
### Enum
- `public enum OtherStatsIndex : byte`
  - `0x01` - PointsFromTrophy
  - `0x02` - PointsFromQuest
  - `0x03` - PointsFromExploratio
  - `0x04` - PointsFromPrestige
- added enum type which is used for all related function calls

### Variables
`public Dictionary<byte, int> OtherStats { get; private set; }`
- added dictionary which keeps track of attribute points gained via prestige/awards/etc
- keys are of OtherStatsIndex type

### StatDistribution.cs
`public StatDistribution(int TotalStats = 0, Dictionary<byte, int> AllocatedStats = null, Dictionary<OtherStatsIndex, int> OtherStats = null)`
- receive total attribute points as argument
- combined constructors; set AllocatedStats/OtherStats to empty dictionary if not passed in
- initialize AllocatedStats/OtherStats if corresponding dictionary passed in

`public void AddTotalStatPoints(int amount, OtherStatsIndex pointSrc);`
- added new function; invoked when gaining attribute points via prestige/awards/etc
- increments TotalStatPoints by amount
- increments corresponding value in OtherStats

### StatPointPacket.cs
`public static Packet WriteTotalStatPoints(Player character);`
- fully implemented total attribute points packet (thanks @nobodylikesaleech )
- send opcode 0x0046 (STAT_POINT), total points mode 0x0, number of sources from which player has received point
- index of each source of attribute points same as above

### Temporary
- added calls to `AddTotalStatPoints(int, byte)` in `StatDistribution()` to demonstrate correctness

Related info can be found here: https://github.com/AlanMorel/MapleServer2/pull/103

### CODE REVIEW

File: MapleServer2/Types/StatDistribution.cs [hunk 1]
```diff
@@ -9,8 +9,9 @@ namespace MapleServer2.Types
     public class StatDistribution
     {
         public int TotalStatPoints { get; private set; }
-        // TODO: implement Dictionary to keep track of points earned from quest, trophy, exploration, prestige
-        // naming convention: PointsFromQuest, PointsFromTrophy, PointsFromExploration, PointsFromPrestige
+        public Dictionary<byte, int> OtherStats { get; private set; }
+        // key: 01 PointsFromTrophy, 02 PointsFromQuest, 03 PointsFromExploration, 04 PointsFromPrestige
```
Comment (AlanMorel) [inline MapleServer2/Types/StatDistribution.cs:13]: Maybe if you turn this into an enum, you don't need this comment, then in lines 26-29, you can use the enum there directly, which also helps self-document the code

File: MapleServer2/Types/StatDistribution.cs [hunk 2]
```diff
@@ -20,19 +21,46 @@ public StatDistribution()
...
+            AddTotalStatPoints(4, 0x4);
         }
 
         public StatDistribution(Dictionary<byte, int> AllocatedStats)
         {
             this.AllocatedStats = AllocatedStats;
+            this.OtherStats = new Dictionary<byte, int>();
+        }
+
+        public StatDistribution(Dictionary<byte, int> AllocatedStats, Dictionary<byte, int> OtherStats)
```
Comment (AlanMorel) [inline MapleServer2/Types/StatDistribution.cs:38]: You can combine these two constructors by making `OtherStats` optional and having it initialize to a blank dictionary if none is passed in

File: MapleServer2/Types/StatDistribution.cs [hunk 3]
```diff
@@ -20,19 +21,46 @@ public StatDistribution()
...
             this.TotalStatPoints += amount;
         }
 
+        // invoked whenever stat points are gained when not leveling up
+        public void AddTotalStatPoints(int amount, byte pointSrc)
+        {
+            if (OtherStats.ContainsKey(pointSrc))
+            {
+                OtherStats[pointSrc] += amount;
+
```
Comment (AlanMorel) [inline MapleServer2/Types/StatDistribution.cs:55]: Unneeded extra blank line

File: MapleServer2/Packets/StatPointPacket.cs [hunk 4]
```diff
@@ -21,7 +21,16 @@ public static Packet WriteTotalStatPoints(Player character)
             // how each stat point was obtained (ie quest, trophy, exploration, prestige)
 
             pWriter.WriteByte((byte) StatPointPacketMode.TotalPoints);
+            // write number of attribute points the character has
             pWriter.WriteInt(character.StatPointDistribution.TotalStatPoints);
+            // write the number of sources from which stat points have been received
+            pWriter.WriteInt(character.StatPointDistribution.OtherStats.Count);
+        
+            foreach (byte pointSrc in character.StatPointDistribution.OtherStats.Keys.ToList())
```
Comment (angelotadeucci) [inline MapleServer2/Packets/StatPointPacket.cs:29]: `.ToList()` is unnecessary.

**Review by angelotadeucci — APPROVED** [review]

File: MapleServer2/Packets/StatPointPacket.cs [hunk 5]
```diff
@@ -26,7 +26,7 @@ public static Packet WriteTotalStatPoints(Player character)
             // write the number of sources from which stat points have been received
             pWriter.WriteInt(character.StatPointDistribution.OtherStats.Count);
         
-            foreach (byte pointSrc in character.StatPointDistribution.OtherStats.Keys.ToList())
+            foreach (MapleServer2.Enums.OtherStatsIndex pointSrc in character.StatPointDistribution.OtherStats.Keys.ToList())
```
Comment (AlanMorel) [inline MapleServer2/Packets/StatPointPacket.cs:29]: Can you import it so you don't need the `MapleServer2.Enums.` part

**Review by AlanMorel — CHANGES_REQUESTED** [review]

**Comment by Andrew-Gan** [comment]
all suggested changes before this message were resolved

**Review by AlanMorel — APPROVED** [review]

**Review by angelotadeucci — APPROVED** [review]

**Pull request merged**